File: Cloude/CloudeShared/Sources/CloudeShared/Messages/ServerMessage.swift
import Foundation

public enum ReplayedEvent: Codable {
    case output(text: String, conversationId: String?, seq: Int)
    case toolCall(name: String, input: String?, toolId: String, parentToolId: String?, conversationId: String?, textPosition: Int?, editInfo: EditInfo?, seq: Int)
    case toolResult(toolId: String, summary: String?, output: String?, conversationId: String?, seq: Int)
    case runStats(durationMs: Int, costUsd: Double, model: String?, conversationId: String?, seq: Int)

    enum CodingKeys: String, CodingKey {
        case kind, text, conversationId, name, input, toolId, parentToolId, textPosition, editInfo, summary, output, durationMs, costUsd, model, seq
    }

    public func encode(to encoder: Encoder) throws {
        var container = encoder.container(keyedBy: CodingKeys.self)
        switch self {
        case .output(let text, let conversationId, let seq):
            try container.encode("output", forKey: .kind)
            try container.encode(text, forKey: .text)
            try container.encodeIfPresent(conversationId, forKey: .conversationId)
            try container.encode(seq, forKey: .seq)
        case .toolCall(let name, let input, let toolId, let parentToolId, let conversationId, let textPosition, let editInfo, let seq):
            try container.encode("tool_call", forKey: .kind)
            try container.encode(name, forKey: .name)
            try container.encodeIfPresent(input, forKey: .input)
            try container.encode(toolId, forKey: .toolId)
            try container.encodeIfPresent(parentToolId, forKey: .parentToolId)
            try container.encodeIfPresent(conversationId, forKey: .conversationId)
            try container.encodeIfPresent(textPosition, forKey: .textPosition)
            try container.encodeIfPresent(editInfo, forKey: .editInfo)
            try container.encode(seq, forKey: .seq)
        case .toolResult(let toolId, let summary, let output, let conversationId, let seq):
            try container.encode("tool_result", forKey: .kind)
            try container.encode(toolId, forKey: .toolId)
            try container.encodeIfPresent(summary, forKey: .summary)
            try container.encodeIfPresent(output, forKey: .output)
            try container.encodeIfPresent(conversationId, forKey: .conversationId)
            try container.encode(seq, forKey: .seq)
        case .runStats(let durationMs, let costUsd, let model, let conversationId, let seq):
            try container.encode("run_stats", forKey: .kind)
            try container.encode(durationMs, forKey: .durationMs)
            try container.encode(costUsd, forKey: .costUsd)
            try container.encodeIfPresent(model, forKey: .model)
            try container.encodeIfPresent(conversationId, forKey: .conversationId)
            try container.encode(seq, forKey: .seq)
        }
    }

    public init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        let kind = try container.decode(String.self, forKey: .kind)
        let seq = try container.decode(Int.self, forKey: .seq)
        switch kind {
        case "output":
            let text = try container.decode(String.self, forKey: .text)
            let conversationId = try container.decodeIfPresent(String.self, forKey: .conversationId)
            self = .output(text: text, conversationId: conversationId, seq: seq)
        case "tool_call":
            let name = try container.decode(String.self, forKey: .name)
            let input = try container.decodeIfPresent(String.self, forKey: .input)
            let toolId = try container.decode(String.self, forKey: .toolId)
            let parentToolId = try container.decodeIfPresent(String.self, forKey: .parentToolId)
            let conversationId = try container.decodeIfPresent(String.self, forKey: .conversationId)
            let textPosition = try container.decodeIfPresent(Int.self, forKey: .textPosition)
            let editInfo = try container.decodeIfPresent(EditInfo.self, forKey: .editInfo)
            self = .toolCall(name: name, input: input, toolId: toolId, parentToolId: parentToolId, conversationId: conversationId, textPosition: textPosition, editInfo: editInfo, seq: seq)
        case "tool_result":
            let toolId = try container.decode(String.self, forKey: .toolId)
            let summary = try container.decodeIfPresent(String.self, forKey: .summary)
            let output = try container.decodeIfPresent(String.self, forKey: .output)
            let conversationId = try container.decodeIfPresent(String.self, forKey: .conversationId)
            self = .toolResult(toolId: toolId, summary: summary, output: output, conversationId: conversationId, seq: seq)
        case "run_stats":
            let durationMs = try container.decode(Int.self, forKey: .durationMs)
            let costUsd = try container.decode(Double.self, forKey: .costUsd)
            let model = try container.decodeIfPresent(String.self, forKey: .model)
            let conversationId = try container.decodeIfPresent(String.self, forKey: .conversationId)
            self = .runStats(durationMs: durationMs, costUsd: costUsd, model: model, conversationId: conversationId, seq: seq)
        default:
            throw DecodingError.dataCorrupted(.init(codingPath: [CodingKeys.kind], debugDescription: "Unknown replayed event kind: \(kind)"))
        }
    }
}

public enum ServerMessage: Codable {
    case output(text: String, conversationId: String?, seq: Int? = nil)
    case status(state: AgentState, conversationId: String?)
    case authRequired
    case authResult(success: Bool, message: String?)
    case error(message: String)
    case directoryListing(path: String, entries: [FileEntry])
    case fileContent(path: String, data: String, mimeType: String, size: Int64, truncated: Bool)
    case sessionId(id: String, conversationId: String?)
    case toolCall(name: String, input: String?, toolId: String, parentToolId: String?, conversationId: String?, textPosition: Int?, editInfo: EditInfo? = nil, seq: Int? = nil)
    case toolResult(toolId: String, summary: String?, output: String?, conversationId: String?, seq: Int? = nil)
    case runStats(durationMs: Int, costUsd: Double, model: String?, conversationId: String?, seq: Int? = nil)
    case resumeFromResponse(sessionId: String, events: [ReplayedEvent], historyOnly: Bool)
    case gitStatusResult(status: GitStatusInfo)
    case gitDiffResult(path: String, diff: String)
    case gitCommitResult(success: Bool, message: String?)
    case gitLogResult(path: String, commits: [GitCommit])
    case transcription(text: String)
    case whisperReady(ready: Bool)
    case defaultWorkingDirectory(path: String)
    case skills([Skill])
    case historySync(sessionId: String, messages: [HistoryMessage])
    case historySyncError(sessionId: String, error: String)
    case fileChunk(path: String, chunkIndex: Int, totalChunks: Int, data: String, mimeType: String, size: Int64)
    case fileThumbnail(path: String, data: String, fullSize: Int64)
    case fileSearchResults(files: [String])
    case messageUUID(uuid: String, conversationId: String?)
    case nameSuggestion(name: String, symbol: String?, conversationId: String)
    case pong(sentAt: Double, serverAt: Double)

    enum CodingKeys: String, CodingKey {
        case type, text, path, diff, state, success, message, entries, data, mimeType, size, truncated, id, sessionId, name, input, status, files, durationMs, costUsd, model, toolId, parentToolId, ready, conversationId, textPosition, symbol, skills, messages, error, chunkIndex, totalChunks, fullSize, uuid, summary, output, filename, editInfo, sentAt, serverAt, commits, seq, events, historyOnly
    }
}

File: android/app/src/main/java/com/cloude/app/Models/ServerMessage.kt
package com.cloude.app.Models

import android.util.Log
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.boolean
import kotlinx.serialization.json.booleanOrNull
import kotlinx.serialization.json.double
import kotlinx.serialization.json.doubleOrNull
import kotlinx.serialization.json.int
import kotlinx.serialization.json.intOrNull
import kotlinx.serialization.json.jsonArray
import kotlinx.serialization.json.jsonObject
import kotlinx.serialization.json.jsonPrimitive
import kotlinx.serialization.json.long

sealed class ServerMessage {
    data class Output(val text: String, val conversationId: String? = null) : ServerMessage()
    data class Status(val state: AgentState, val conversationId: String? = null) : ServerMessage()
    data object AuthRequired : ServerMessage()
    data class AuthResult(val success: Boolean, val message: String? = null) : ServerMessage()
    data class Error(val message: String) : ServerMessage()
    data class DirectoryListing(val path: String, val entries: List<FileEntry>) : ServerMessage()
    data class FileContent(val path: String, val data: String, val mimeType: String, val size: Long, val truncated: Boolean = false) : ServerMessage()
    data class SessionId(val id: String, val conversationId: String? = null) : ServerMessage()
    data class MissedResponse(val sessionId: String, val text: String, val completedAt: Double, val toolCalls: List<StoredToolCall> = emptyList()) : ServerMessage()
    data class NoMissedResponse(val sessionId: String) : ServerMessage()
    data class ToolCall(val name: String, val input: String? = null, val toolId: String, val parentToolId: String? = null, val conversationId: String? = null, val textPosition: Int? = null, val editInfo: EditInfo? = null) : ServerMessage()
    data class ToolResult(val toolId: String, val summary: String? = null, val output: String? = null, val conversationId: String? = null) : ServerMessage()
    data class RunStats(val durationMs: Int, val costUsd: Double, val model: String? = null, val conversationId: String? = null) : ServerMessage()
    data class GitStatusResult(val status: GitStatusInfo) : ServerMessage()
    data class GitDiffResult(val path: String, val diff: String) : ServerMessage()
    data class GitCommitResult(val success: Boolean, val message: String? = null) : ServerMessage()
    data class GitLogResult(val path: String, val commits: List<GitCommit>) : ServerMessage()
    data class Transcription(val text: String) : ServerMessage()
    data class WhisperReady(val ready: Boolean) : ServerMessage()
    data class DefaultWorkingDirectory(val path: String) : ServerMessage()
    data class Skills(val skills: List<Skill>) : ServerMessage()
    data class HistorySync(val sessionId: String, val messages: List<HistoryMessage>) : ServerMessage()
    data class HistorySyncError(val sessionId: String, val error: String) : ServerMessage()
    data class FileChunk(val path: String, val chunkIndex: Int, val totalChunks: Int, val data: String, val mimeType: String, val size: Long) : ServerMessage()
    data class FileThumbnail(val path: String, val data: String, val fullSize: Long) : ServerMessage()
    data class FileSearchResults(val files: List<String>) : ServerMessage()
    data class MessageUUID(val uuid: String, val conversationId: String? = null) : ServerMessage()
    data class NameSuggestion(val name: String, val symbol: String? = null, val conversationId: String) : ServerMessage()

    companion object {
        private val json = Json { ignoreUnknownKeys = true; isLenient = true }

        fun decode(text: String): ServerMessage? = try {
            doDecode(text)
        } catch (e: Exception) {
            Log.e("Cloude", "ServerMessage decode error: ${e.message} for: ${text.take(200)}")
            null
        }

        private fun doDecode(text: String): ServerMessage? {
            val obj = json.parseToJsonElement(text).jsonObject
            val type = obj["type"]?.jsonPrimitive?.content ?: return null

            return when (type) {
                "output" -> Output(
                    text = obj.str("text") ?: "",
                    conversationId = obj.strOrNull("conversationId")
                )
                "status" -> Status(
                    state = when (obj.str("state")) {
                        "running" -> AgentState.running
                        "compacting" -> AgentState.compacting
                        else -> AgentState.idle
                    },
                    conversationId = obj.strOrNull("conversationId")
                )
                "auth_required" -> AuthRequired
                "auth_result" -> AuthResult(
                    success = obj["success"]?.jsonPrimitive?.boolean ?: false,
                    message = obj.strOrNull("message")
                )
                "error" -> Error(message = obj.str("message") ?: "Unknown error")
                "session_id" -> SessionId(
                    id = obj.str("id") ?: "",
                    conversationId = obj.strOrNull("conversationId")
                )
                "tool_call" -> ToolCall(
                    name = obj.str("name") ?: "",
                    input = obj.strOrNull("input"),
                    toolId = obj.str("toolId") ?: "",
                    parentToolId = obj.strOrNull("parentToolId"),
                    conversationId = obj.strOrNull("conversationId"),
                    textPosition = obj["textPosition"]?.jsonPrimitive?.intOrNull,
                    editInfo = obj["editInfo"]?.let { json.decodeFromJsonElement(EditInfo.serializer(), it) }
                )
                "tool_result" -> ToolResult(
                    toolId = obj.str("toolId") ?: "",
                    summary = obj.strOrNull("summary"),
                    output = obj.strOrNull("output"),
                    conversationId = obj.strOrNull("conversationId")
                )
                "run_stats" -> RunStats(
                    durationMs = obj["durationMs"]?.jsonPrimitive?.int ?: 0,
                    costUsd = obj["costUsd"]?.jsonPrimitive?.double ?: 0.0,
                    model = obj.strOrNull("model"),
                    conversationId = obj.strOrNull("conversationId")
                )
                "message_uuid" -> MessageUUID(
                    uuid = obj.str("uuid") ?: "",
                    conversationId = obj.strOrNull("conversationId")
                )
                "missed_response" -> MissedResponse(
                    sessionId = obj.str("sessionId") ?: "",
                    text = obj.str("text") ?: "",
                    completedAt = obj["completedAt"]?.jsonPrimitive?.doubleOrNull ?: 0.0,
                    toolCalls = obj["toolCalls"]?.jsonArray?.map { json.decodeFromJsonElement(StoredToolCall.serializer(), it) } ?: emptyList()
                )
                "no_missed_response" -> NoMissedResponse(sessionId = obj.str("sessionId") ?: "")
                "directory_listing" -> DirectoryListing(
                    path = obj.str("path") ?: "",
                    entries = obj["entries"]?.jsonArray?.map { json.decodeFromJsonElement(FileEntry.serializer(), it) } ?: emptyList()
                )
                "file_content" -> FileContent(
                    path = obj.str("path") ?: "",
                    data = obj.str("data") ?: "",
                    mimeType = obj.str("mimeType") ?: "",
                    size = obj["size"]?.jsonPrimitive?.long ?: 0,
                    truncated = obj["truncated"]?.jsonPrimitive?.booleanOrNull ?: false
                )
                "git_status_result" -> GitStatusResult(
                    status = json.decodeFromJsonElement(GitStatusInfo.serializer(), obj["status"]?.jsonObject ?: return null)
                )
                "git_diff_result" -> GitDiffResult(
                    path = obj.str("path") ?: "",
                    diff = obj.str("diff") ?: ""
                )
                "git_commit_result" -> GitCommitResult(
                    success = obj["success"]?.jsonPrimitive?.boolean ?: false,
                    message = obj.strOrNull("message")
                )
                "git_log_result" -> GitLogResult(
                    path = obj.str("path") ?: "",
                    commits = obj["commits"]?.jsonArray?.map { json.decodeFromJsonElement(GitCommit.serializer(), it) } ?: emptyList()
                )
                "transcription" -> Transcription(text = obj.str("text") ?: "")
                "whisper_ready" -> WhisperReady(ready = obj["ready"]?.jsonPrimitive?.boolean ?: false)
                "default_working_directory" -> DefaultWorkingDirectory(path = obj.str("path") ?: "")
                "skills" -> Skills(
                    skills = obj["skills"]?.jsonArray?.map { json.decodeFromJsonElement(Skill.serializer(), it) } ?: emptyList()
                )
                "history_sync" -> HistorySync(
                    sessionId = obj.str("sessionId") ?: "",
                    messages = obj["messages"]?.jsonArray?.map { json.decodeFromJsonElement(HistoryMessage.serializer(), it) } ?: emptyList()
                )
                "history_sync_error" -> HistorySyncError(
                    sessionId = obj.str("sessionId") ?: "",
                    error = obj.str("error") ?: ""
                )
                "file_chunk" -> FileChunk(
                    path = obj.str("path") ?: "",
                    chunkIndex = obj["chunkIndex"]?.jsonPrimitive?.int ?: 0,
                    totalChunks = obj["totalChunks"]?.jsonPrimitive?.int ?: 0,
                    data = obj.str("data") ?: "",
                    mimeType = obj.str("mimeType") ?: "",
                    size = obj["size"]?.jsonPrimitive?.long ?: 0
                )
                "file_thumbnail" -> FileThumbnail(
                    path = obj.str("path") ?: "",
                    data = obj.str("data") ?: "",
                    fullSize = obj["fullSize"]?.jsonPrimitive?.long ?: 0
                )
                "file_search_results" -> FileSearchResults(
                    files = obj["files"]?.jsonArray?.map { it.jsonPrimitive.content } ?: emptyList()
                )
                "name_suggestion" -> NameSuggestion(
                    name = obj.str("name") ?: "",
                    symbol = obj.strOrNull("symbol"),
                    conversationId = obj.str("conversationId") ?: ""
                )
                "team_created", "teammate_spawned", "teammate_update", "team_deleted" -> null
                else -> null
            }
        }

        private fun JsonObject.str(key: String): String? =
            this[key]?.jsonPrimitive?.content

        private fun JsonObject.strOrNull(key: String): String? =
            this[key]?.jsonPrimitive?.takeIf { it.isString }?.content
    }
}
